src/once_cell.rs

```diff
@@ -0,0 +1,431 @@
+                    // were just waiting, so there must've been waiters.
+                    State::Uninit => unreachable!(),
+                    // The previous initializer failed to initialize the cell (it panicked or
+                    // errored). The job has now been passed down to us.
+                    State::Initializing => {}
+                    State::Initialized { value } => return Ok(value),
+                }
+            }
+            State::Initialized { value } => return Ok(value),
+        }
+
+        // This guard's Drop implementation is activated when `f` errors or panics and performs the
+        // necessary cleanup of notifying others that we have failed.
+        struct Guard<'once_cell, T>(&'once_cell OnceCell<T>);
+        impl<T> Drop for Guard<'_, T> {
+            fn drop(&mut self) {
+                // We failed to initialize, so attempt to pass the job of initialization onto
+                // the next waiter.
+                if self.0.waiters.borrow().wake_one(()).is_err() {
+                    // Having failed that, no-one is initializing the cell, so we must set its
+                    // state back to `Uninit`.
+                    unsafe { *self.0.state.get() = State::Uninit };
+                }
+            }
+        }
+
+        let guard = Guard(self);
+
+        let value = f().await?;
+        unsafe { *self.state.get() = State::Initialized { value } };
+
+        // Disarm the guard
+        mem::forget(guard);
+
+        let mut waiters = self.waiters.borrow();
+        while waiters.wake_one(()).is_ok() {}
+
+        Ok(self.get().unwrap())
+    }
+
+    /// Take the value out of the `OnceCell`, leaving it in an uninitialized state.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use unsync::once_cell::OnceCell;
+    ///
+    /// let mut cell = OnceCell::new();
+    /// assert_eq!(cell.take(), None);
+    ///
+    /// cell.set(10).unwrap();
+    /// assert_eq!(cell.take(), Some(10));
+    /// assert_eq!(cell.take(), None);
+    /// ```
+    pub fn take(&mut self) -> Option<T> {
+        mem::take(self).into_inner()
+    }
+
+    /// Consume this `OnceCell` and return the inner value, if there was any.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use unsync::once_cell::OnceCell;
+    ///
+    /// assert_eq!(<OnceCell<i32>>::new().into_inner(), None);
+    /// assert_eq!(<OnceCell<i32>>::from(10).into_inner(), Some(10));
+    /// ```
+    pub fn into_inner(self) -> Option<T> {
+        match self.state.into_inner() {
+            State::Uninit | State::Initializing => None,
+            State::Initialized { value } => Some(value),
+        }
+    }
+}
+
+impl<T: Debug> Debug for OnceCell<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut s = f.debug_struct("OnceCell");
+        if let Some(inner) = self.get() {
+            s.field("value", inner);
```

Comment:
Might as well debug the `state` field as-is. We can then have `#[derive(Debug)]`.